### ISSUE TITLE
cmd/compile: detect call trees that can be marked NOSPLIT

### ISSUE THREAD

**Issue by josharian** [golang/go#17023]
CL 24814 contains a minor optimization for 386/amd64: Ignore a few common runtime functions that take no arguments when deciding whether a function is a leaf function and is thus eligible for being marked NOSPLIT and thus having no function prolog.

Keith suggested some ways during code review to make this analysis much more powerful. This issue is a reminder to investigate those suggestions.

**Comment by martisch** [comment]
link to code review: https://go-review.googlesource.com/#/c/24814/

**Comment by cherrymui** [comment]
Indeed I've been thinking about something similar at link time, where we know the full call tree (at least for internal static linking), initially motivated for reducing binary size. But I am not sure about the complexity vs. gain (indirect calls are tricky). I may do some experiment.

**Comment by josharian** [comment]
I'd be surprised if this had a significant impact on binary size. The prologues aren't that big.

**Comment by cherrymui** [comment]
I agree that the benefit may be quite small. So it might not be worth doing...

**Comment by josharian** [comment]
There could still be nice perf improvements. The CL above picks up a percent or two in the encoding/binary benchmarks.
